RFCs/0035-partiql-window-function.md

```diff
@@ -0,0 +1,1334 @@
+--   },
+--   {
+--     'current_month': 9, 'ticker': 'GOOG',
+--     'current_month_average': 96.15, 'previous_month_avg': NULL
+--   }
+-- >>
+```
+
+Example 3.3: Outer Order By
+
+```sql
+-- query
+SELECT s.trade_date as trade_date,
+       s.ticker as ticker,
+       s.price as current_price,
+       lag(s.price) OVER (PARTITION BY s.ticker ORDER BY s.trade_date) as previous_price
+FROM stock as s
+ORDER BY s.trade_date DESC
+
+-- Equivalent PartiQL Core query
+-- See Example 1.4 for the Binding Tuples created by WINDOWED clause 
+SELECT value {
+    'date' : s.trade_date,
+    'ticker' : s.ticker,
+    'previous_price': 
+        CASE WHEN ws_1_pos - 1 >= 0 THEN
+            CAST (SELECT VALUE s.price FROM ws_1_partition AT idx WHERE idx = ws_1_pos - 1 AS LIST)[0]
+        ELSE NULL END
+    }
+FROM stock as s
+WINDOWED
+    (PARTITION BY s.ticker ORDER BY s.trade_date) 
+        PARTITION AS ws_1_partition AT ws_1_pos
+ORDER BY s.trade_date DESC
+
+-- result:
+-- [
+--   {
+--     'trade_date': `2022-10-03`, 'ticker': 'AMZN',
+--     'current_price': 115.88, 'previous_price': 113.00
+--   },
+--   {
+--     'trade_date': `2022-09-30`, 'ticker': 'AMZN',
+--     'current_price': 113.00, 'previous_price': NULL
+--   },
+--   {
+--     'trade_date': `2022-09-30`, 'ticker': 'GOOG',
+--     'current_price': 96.15, 'previous_price': NULL
+--   }
+-- ]
+```
+
+Notice here the output is a list, whereas the results in previous query return a bag. This is because the ORDER BY sub-clause only order the partition.
+
+Example 3.5: Missing Binding
+
+It is worth to distinguish between `out of window partition` and `no such binding.`
+
+Consider:
+
+```sql
+-- query
+SELECT 
+       lag(s.a, 1, 'Out Of Partition') 
+            OVER (PARTITION BY s.ticker ORDER BY s.trade_date) as previous_a
+FROM stock as s
+
+-- Equivalent PartiQL Core query
+SELECT value {
+    'previous_a': 
+        CASE WHEN ws_1_pos - 1 >= 0 THEN
+            CAST (SELECT VALUE s.a FROM ws_1_partition AT idx WHERE idx = ws_1_pos - 1 AS LIST)[0]
+        ELSE 'Out Of Partition' END
+    }
+FROM stock as s
+WINDOWED
+    (PARTITION BY s.ticker ORDER BY s.trade_date) 
+        PARTITION AS ws_1_partition AT ws_1_pos
+
+-- result
+-- <<
+--  {
+--    'previous_a': 'Out Of Partition'
+--  },
+--  {},
+--  {
+--    'previous_a': 'Out Of Partition'
+--  }
+>>
+```
+
+The first row returns `{ 'previous_a': 'Out of Partition' }` because current row is the first row in partition, and `lag(sp.a)` tries to access the row before, which is out of the window partition. Therefore, it returns the default value.
+The second row returns an empty struct `{}` , this is because the current row is now the second row, and `lag(sp.a)` essentially evaluates `sp.a` over the binding tuple `{ 'trade_date': `2022-09-30`, 'ticker': 'AMZN', 'price': 113.00}`. Since there is no binding name `a` in the binding tuple, the query returns `missing` in permissive mode.
```

Comment:
```suggestion
The first row returns `{ 'previous_a': 'Out of Partition' }` because current row is the first row in partition, and `lag(sp.a)` tries to access the row before, which is out of the window partition. Therefore, it returns the default value.
The second row returns an empty struct `{}` , this is because the current row is now the second row, and `lag(sp.a)` essentially evaluates `sp.a` over the binding tuple `{ 'trade_date': `2022-09-30`, 'ticker': 'AMZN', 'price': 113.00}`. Since there is no binding name `a` in the binding tuple, the query returns `missing` in permissive mode as outlined in [PartiQL specification 2019](https://partiql.org/assets/PartiQL-Specification.pdf).
```